Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,365 @@
+        var isClosed: Atomic<Bool>
+        /// The curently opened, unleased connections in this connection provider
+        var availableConnections: CircularBuffer<Connection> = .init(initialCapacity: 8) {
+            didSet {
+                self.closeIfNeeded()
+            }
+        }
+
+        /// Consumers that weren't able to get a new connection without exceeding
+        /// `maximumConcurrentConnections` get a `Future<Connection>`
+        /// whose associated promise is stored in `Waiter`. The promise is completed
+        /// as soon as possible by the provider, in FIFO order.
+        var waiters: CircularBuffer<Waiter> = .init()
+
+        var leased: Int = 0 {
+            didSet {
+                self.closeIfNeeded()
+            }
+        }
+
+        let maximumConcurrentConnections: Int = 8
+
+        let lock = Lock()
+
+        let parentPool: ConnectionPool
+
+        init(group: EventLoopGroup, key: ConnectionPool.Key, configuration: HTTPClient.Configuration, initialConnection: Connection, parentPool: ConnectionPool) {
+            self.loopGroup = group
+            self.configuration = configuration
+            self.key = key
+            self.availableConnections.append(initialConnection)
+            self.parentPool = parentPool
+            self.isClosed = Atomic(value: false)
+            self.configureClose(of: initialConnection)
+        }
+
+        deinit {
+            assert(availableConnections.isEmpty, "Available connections should be empty before deinit")
+            assert(leased == 0, "All leased connections should have been returned before deinit")
+        }
+
+        /// Get a new connection from this provider
+        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
+            return self.lock.withLock {
+                if leased < maximumConcurrentConnections {
+                    leased += 1
+                    if let connection = availableConnections.popFirst(), connection.channel.isActive {
+                        connection.isLeased = true
+                        return self.loopGroup.next().makeSucceededFuture(connection)
+                    } else {
+                        return self.makeConnection()
+                    }
+                } else {
+                    let promise = self.loopGroup.next().makePromise(of: Connection.self)
+                    self.waiters.append(Waiter(promise: promise, preference: preference))
+                    return promise.futureResult
+                }
+            }
+        }
+
+        private func configureClose(of connection: Connection) {
+            connection.channel.closeFuture.whenComplete { _ in
+                if !connection.isLeased {
+                    if let indexToRemove = self.availableConnections.firstIndex(where: { $0 === connection }) {
+                        self.availableConnections.swapAt(self.availableConnections.startIndex, indexToRemove)
```

Comment:
there's a threading bug here. `connection.channel.eventLoop` where the `whenComplete` will run is essentially a random thread but you're accessing `self.availableConnections` from there.

Reply:
did you run the tests under TSan yet?

Reply:
@weissi I did and it showed multiple issues indeed